usr/lib/api/policy.c

```diff
@@ -0,0 +1,1728 @@
+            /* Fallthrough */
+        case CKM_ECDSA_SHA1:
+            /* Fallthrough */
+        case CKM_SHA1_RSA_PKCS:
+            /* Fallthrough */
+        case CKM_SHA1_RSA_PKCS_PSS:
+            /* Fallthrough */
+        case CKM_SHA1_RSA_X9_31:
+            *ssize = MIN(s->siglen, 160);
+            break;
+        case CKM_MD5_RSA_PKCS:
+            *ssize = MIN(s->siglen, 128);
+            break;
+        case CKM_ECDSA_SHA224:
+            /* Fallthrough */
+        case CKM_SHA224_RSA_PKCS:
+            /* Fallthrough */
+        case CKM_SHA224_RSA_PKCS_PSS:
+            *ssize = MIN(s->siglen, 224);
+            break;
+        case CKM_ECDSA_SHA256:
+            /* Fallthrough */
+        case CKM_SHA256_RSA_PKCS:
+            /* Fallthrough */
+        case CKM_SHA256_RSA_PKCS_PSS:
+            *ssize = MIN(s->siglen, 256);
+            break;
+        case CKM_ECDSA_SHA384:
+            /* Fallthrough */
+        case CKM_SHA384_RSA_PKCS:
+            /* Fallthrough */
+        case CKM_SHA384_RSA_PKCS_PSS:
+            *ssize = MIN(s->siglen, 384);
+            break;
+        case CKM_ECDSA_SHA512:
+            /* Fallthrough */
+        case CKM_SHA512_RSA_PKCS:
+            /* Fallthrough */
+        case CKM_SHA512_RSA_PKCS_PSS:
+            *ssize = MIN(s->siglen, 512);
+            break;
+        default:
+            return CKR_FUNCTION_FAILED;
+        }
+        return CKR_OK;
+    }
+    if (col->outputsize == MC_INFORMATION_UNAVAILABLE)
+        return CKR_FUNCTION_FAILED;
+    *ssize = col->outputsize * 8;
+    return CKR_OK;
+}
+
+static inline CK_RV policy_is_mgf_allowed(struct policy_private *pp,
+                                          CK_RSA_PKCS_MGF_TYPE mgf)
+{
+    if (mgf > CKG_VENDOR_DEFINED) {
+        if (pp->allowedvendormgfs & (1u << (mgf - CKG_VENDOR_DEFINED - 1)))
+            return CKR_OK;
+    } else {
+        if (pp->allowedmgfs & (1u << mgf))
+            return CKR_OK;
+    }
+    TRACE_WARNING("POLICY VIOLATION: mgf not allowed: 0x%lx\n", mgf);
+    return CKR_FUNCTION_FAILED;
+}
+
+static inline CK_RV policy_is_kdf_allowed(struct policy_private *pp,
+                                          CK_ULONG kdf)
+{
+    if (pp->allowedkdfs & (1u << kdf))
+        return CKR_OK;
+    TRACE_WARNING("POLICY VIOLATION: kdf not allowed: 0x%lx\n", kdf);
+    return CKR_FUNCTION_FAILED;
+}
+
+static inline CK_RV policy_is_prf_allowed(struct policy_private *pp,
+                                          CK_ULONG val)
+{
+    if (pp->allowedprfs & (1u << val))
+        return CKR_OK;
+    TRACE_WARNING("POLICY VIOLATION: prf not allowed: 0x%lx\n", val);
+    return CKR_FUNCTION_FAILED;
+}
+
+static CK_RV policy_is_key_allowed_i(struct policy_private *pp,
+                                     struct objstrength *s)
+{
+    /* 1. Check strength */
+    if (s->strength > pp->minstrengthidx) {
+        TRACE_WARNING("POLICY VIOLATION: Key does not satisfy minimal strength constraint\n");
```

Comment:
Its hard to know for which key / object such trace message is issued:

Example from my token init output:
```
11/03/2021 10:01:36 3125 [usr/lib/api/policy.c:441 api] WARN: POLICY VIOLATION: Key does not satisfy minimal strength constraint
11/03/2021 10:01:36 3125 [usr/lib/api/policy.c:441 api] WARN: POLICY VIOLATION: Key does not satisfy minimal strength constraint
11/03/2021 10:01:36 3125 [usr/lib/api/policy.c:441 api] WARN: POLICY VIOLATION: Key does not satisfy minimal strength constraint
11/03/2021 10:01:36 3125 [usr/lib/api/policy.c:441 api] WARN: POLICY VIOLATION: Key does not satisfy minimal strength constraint
11/03/2021 10:01:36 3125 [usr/lib/api/policy.c:441 api] WARN: POLICY VIOLATION: Key does not satisfy minimal strength constraint
11/03/2021 10:01:36 3125 [usr/lib/api/policy.c:441 api] WARN: POLICY VIOLATION: Key does not satisfy minimal strength constraint
```
However, I also don't know how to make this better.... So you may just leave it as is.

Reply:
Since at that point I have no other information, the caller would have to provide additional information.  I cannot do anything here (except maybe print the strength that was determined, but this does not really help).